File: packages/modules/api/go/src/api/telegram.go
package api

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"net/url"
)

const baseURL = "https://api.telegram.org/bot"

type ParseMode string

const (
	ParseModeHTML     ParseMode = "html"
	ParseModeMarkdown ParseMode = "markdown"
)

type SetWebhookResponse struct {
	OK          bool   `json:"ok"`
	Result      bool   `json:"result"`
	Description string `json:"description"`
}

type DeleteWebhookResponse struct {
	OK          bool   `json:"ok"`
	Result      bool   `json:"result"`
	Description string `json:"description"`
}

type WebhookInfo struct {
	OK     bool `json:"ok"`
	Result struct {
		URL                  string `json:"url"`
		HasCustomCertificate bool   `json:"has_custom_certificate"`
		PendingUpdateCount   int    `json:"pending_update_count"`
	} `json:"result"`
}

func post(urlStr string, requestBody interface{}) (*http.Response, error) {
	jsonData, err := json.Marshal(requestBody)
	if err != nil {
		return nil, err
	}
	return http.Post(urlStr, "application/json", bytes.NewBuffer(jsonData))
}

func SendMessage(token string, chatID int, message string, parseMode ParseMode) error {
	if token == "" {
		return errors.New("invalid token")
	}
	if chatID == 0 {
		return errors.New("invalid chatID")
	}
	if message == "" {
		return errors.New("invalid message")
	}

	u, _ := url.Parse(fmt.Sprintf("%s%s/sendMessage", baseURL, token))
	params := u.Query()
	params.Set("chat_id", fmt.Sprintf("%d", chatID))
	params.Set("text", message)
	params.Set("parse_mode", string(parseMode))
	u.RawQuery = params.Encode()

	resp, err := http.Get(u.String())
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("api request failed with status: %d", resp.StatusCode)
	}
	return nil
}

func SetWebhook(token, webhookURL string) (*SetWebhookResponse, error) {
	if token == "" {
		return nil, errors.New("invalid token")
	}
	if webhookURL == "" {
		return nil, errors.New("invalid url")
	}

	resp, err := post(fmt.Sprintf("%s%s/setWebhook", baseURL, token), map[string]string{"url": webhookURL})
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result SetWebhookResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return &result, nil
}

func DeleteWebhook(token, webhookURL string) (*DeleteWebhookResponse, error) {
	if token == "" {
		return nil, errors.New("invalid token")
	}
	if webhookURL == "" {
		return nil, errors.New("invalid url")
	}

	resp, err := post(fmt.Sprintf("%s%s/deleteWebhook", baseURL, token), map[string]string{"url": webhookURL})
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var result DeleteWebhookResponse
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}
	return &result, nil
}

func GetWebhookInfo(token string) (*WebhookInfo, error) {
	if token == "" {
		return nil, errors.New("invalid token")
	}

	resp, err := http.Get(fmt.Sprintf("%s%s/getWebhookInfo", baseURL, token))
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var info WebhookInfo
	if err := json.NewDecoder(resp.Body).Decode(&info); err != nil {
		return nil, err
	}
	return &info, nil
}

File: packages/cli/hieudoanm/rust/src/services/telegram.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};

const TELEGRAM_API: &str = "https://api.telegram.org/bot";

#[derive(Debug, Serialize)]
pub struct SendMessageParams {
    pub chat_id: String,
    pub text: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub parse_mode: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub disable_web_page_preview: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct TelegramResponse {
    pub ok: bool,
    pub description: Option<String>,
}

pub fn send_message(token: &str, chat_id: &str, text: &str) -> Result<TelegramResponse> {
    let url = format!("{}{}/sendMessage", TELEGRAM_API, token);
    let params = SendMessageParams {
        chat_id: chat_id.to_string(),
        text: text.to_string(),
        parse_mode: Some("HTML".into()),
        disable_web_page_preview: Some(true),
    };

    let client = reqwest::blocking::Client::new();
    let resp = client
        .post(&url)
        .json(&params)
        .send()
        .context("telegram API request failed")?;

    let result: TelegramResponse = resp.json().context("failed to parse telegram response")?;
    Ok(result)
}

pub fn set_webhook(token: &str, webhook_url: &str) -> Result<TelegramResponse> {
    let url = format!("{}{}/setWebhook?url={}", TELEGRAM_API, token, urlencoding(webhook_url));

    let client = reqwest::blocking::Client::new();
    let resp = client
        .get(&url)
        .send()
        .context("telegram webhook request failed")?;

    let result: TelegramResponse = resp.json().context("failed to parse telegram response")?;
    Ok(result)
}

pub fn delete_webhook(token: &str) -> Result<TelegramResponse> {
    let url = format!("{}{}/deleteWebhook", TELEGRAM_API, token);

    let client = reqwest::blocking::Client::new();
    let resp = client
        .get(&url)
        .send()
        .context("telegram delete webhook request failed")?;

    let result: TelegramResponse = resp.json().context("failed to parse telegram response")?;
    Ok(result)
}

pub fn get_webhook_info(token: &str) -> Result<serde_json::Value> {
    let url = format!("{}{}/getWebhookInfo", TELEGRAM_API, token);

    let client = reqwest::blocking::Client::new();
    let resp = client
        .get(&url)
        .send()
        .context("telegram webhook info request failed")?;

    let result: serde_json::Value = resp.json().context("failed to parse telegram response")?;
    Ok(result)
}

fn urlencoding(s: &str) -> String {
    url::form_urlencoded::byte_serialize(s.as_bytes()).collect()
}
